src/share/termtypes/termcap.5
.\"	$MirOS$
.\"	$OpenBSD: termcap.5,v 1.17 2005/03/12 12:21:08 jmc Exp $
.\"
.\" Copyright (c) 1985, 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)termcap.5	6.11 (Berkeley) 3/6/93
.\"
.Dd March 6, 1993
.Dt TERMCAP 5
.Os
.Sh NAME
.Nm termcap
.Nd terminal capability database
.Sh SYNOPSIS
.Nm termcap
.Sh DESCRIPTION
The
.Nm
file
is a database describing terminals, used, for example, by
.Xr \&vi 1
and
.Xr curses 3 .
Terminals are described in
.Nm
by giving a set of capabilities that they have and by describing
how operations are performed.
Padding requirements and initialization sequences
are included in
.Nm termcap .
.Pp
Entries in
.Nm
consist of a number of `:'-separated fields.
The first entry for each terminal gives the names that are known for the
terminal, separated by `|' characters.
The first name given is the most common abbreviation for the terminal.
The last name given should be a long name fully identifying the terminal,
and all others are understood as synonyms for the terminal name.
All names but the last should be in lower case and contain no blanks;
the last name may well contain upper case characters and blanks for
readability.
.Pp
Terminal names (except for the last, verbose entry)
should be chosen using the following conventions.
The particular piece of hardware making up the terminal
should have a root name chosen, thus
.Dq hp2621
This name should not contain hyphens.
Modes that the hardware can be in
or user preferences
should be indicated by appending a hyphen and an indicator of the mode.
Therefore, a
.Dq vt100
in 132-column mode would be
.Dq vt100-w .
.Pp
The following suffixes should be used where possible:
.Bd -filled
.Bl -column indent "With automatic margins (usually default)xx"
.Sy Suffix	Meaning	Example
-w	Wide mode (more than 80 columns).	vt100-w
-am	With automatic margins (usually default).	vt100-am
-nam	Without automatic margins.	vt100-nam
-\fIn\fR	Number of lines on screen.	aaa-60
-na	No arrow keys (leave them in local).	concept100-na
-\fInp\fR	Number of pages of memory.	concept100-4p
-rv	Reverse video.	concept100-rv
.El
.Ed
.Sh CAPABILITIES
The characters in the
.Em Notes
function
field in the table have the following meanings
(more than one may apply to a capability):
.Bd -unfilled
N	Indicates numeric parameter(s).
P	Indicates that padding may be specified.
*	Indicates that padding may be based on the number of lines affected.
o	Indicates capability is obsolete.
.Ed
.Pp
.Dq Obsolete
capabilities have no
.Em terminfo
equivalents, since they were considered useless,
or are subsumed by other capabilities.
New software should not rely on them at all.
.Bl -column indent indent indent
.Sy Name	Type	Notes	Description
.It "ae	str	(P)	End alternate character set."
.It "AL	str	(NP*)	Add"
.Em n
new blank lines
.It "al	str	(P*)	Add new blank line."
.It "am	bool		Terminal has automatic margins."
.It "as	str	(P)	Start alternate character set."
.It "bc	str	(o)	Backspace if not"
.Sy \&^H .
.It "bl	str	(P)	Audible signal (bell)."
.It "bs	bool	(o)	Terminal can backspace with"
.Sy \&^H .
.It "bt	str	(P)	Back tab."
.It "bw	bool	" Ta Sy \&le
(backspace) wraps from column 0 to last column.
.It "CC	str		Terminal settable command character in prototype."
.It "cd	str	(P*)	Clear to end of display."
.It "ce	str	(P)	Clear to end of line."
.It "ch	str	(NP)	Set cursor column (horizontal position)."
.It "cl	str	(P*)	Clear screen and home cursor."
.It "CM	str	(NP)	Memory-relative cursor addressing."
.It "cm	str	(NP)	Screen-relative cursor motion."
.It "co	num		Number of columns in a line (see"
.Sx BUGS
section below).
.It "cr	str	(P)	Carriage return."
.It "cs	str	(NP)	Change scrolling region (VT100)."
.It "ct	str	(P)	Clear all tab stops."
.It "cv	str	(NP)	Set cursor row (vertical position)."
.It "da	bool		Display may be retained above the screen."
.It "dB	num	(o)	Milliseconds"
.Sy \&bs
delay needed (default 0).
.It "db	bool		Display may be retained below the screen."
.It "DC	str	(NP*)	Delete"
.Em n
characters.
.It "dC	num	(o)	Milliseconds"
.Sy \&cr
delay needed (default 0).
.It "dc	str	(P*)	Delete character."
.It "dF	num	(o)	Milliseconds"
.Sy \&ff
delay needed (default 0).
.It "DL	str	(NP*)	Delete"
.Ar n
lines.
.It "dl	str	(P*)	Delete line."
.It "dm	str		Enter delete mode."
.It "dN	num	(o)	Milliseconds"
.Sy \&nl
delay needed (default 0).
.It "DO	str	(NP*)	Move cursor down:
.Ar n
lines.
.It "do	str		Down one line."
.It "ds	str		Disable status line."
.It "dT	num	(o)	Milliseconds of horizontal tab delay needed (default 0)."
.It "dV	num	(o)	Milliseconds of vertical tab delay needed (default 0)."
.It "ec	str	(NP)	Erase"
.Ar n
characters.
.It "ed	str		End delete mode."
.It "ei	str		End insert mode."
.It "eo	bool		Can erase overstrikes with a blank."
.It "EP	bool	(o)	Even parity."
.It "es	bool		Escape can be used on the status line."
.It "ff	str	(P*)	Hardcopy terminal page eject."
.It "fs	str		Return from status line."
.It "gn	bool		Generic line type, for example dialup, switch)."
.It "hc	bool		Hardcopy terminal."
.It "HD	bool	(o)	Half-duplex."
.It "hd	str		Half-line down (forward 1/2 linefeed)."
.It "ho	str	(P)	Home cursor."
.It "hs	bool		Has extra"
.Dq status line .
.It "hu	str		Half-line up (reverse 1/2 linefeed)."
.It "hz	bool		Cannot print ``~'' (Hazeltine)."
.It "i1-i3	str		Terminal initialization strings"
.Pf ( Xr terminfo
only)
.It "IC	str	(NP*)	Insert"
.Ar n
blank characters.
.It "ic	str	(P*)	Insert character."
.It "if	str		Name of file containing initialization string."
.It "im	str		Enter insert mode."
.It "in	bool		Insert mode distinguishes nulls."
.It "iP	str		Pathname of program for initialization"
.Pf ( Xr terminfo
only).
.It "ip	str	(P*)	Insert pad after character inserted."
.It "is	str		Terminal initialization string"
.Pf ( Nm termcap
only).
.It "it	num		Tabs initially every"
.Ar n
positions.
.It "K1	str		Sent by keypad upper left."
.It "K2	str		Sent by keypad center."
.It "K3	str		Sent by keypad upper right."
.It "K4	str		Sent by keypad lower left."
.It "K5	str		Sent by keypad lower right."
.It "k0-k9	str		Sent by function keys 0-9."
.It "kA	str		Sent by insert-line key."
.It "ka	str		Sent by clear-all-tabs key."
.It "kb	str		Sent by backspace key."
.It "kC	str		Sent by clear-screen or erase key."
.It "kD	str		Sent by delete-character key."
.It "kd	str		Sent by down-arrow key."
.It "kE	str		Sent by clear-to-end-of-line key."
.It "ke	str		Out of"
.Dq keypad transmit
mode.
.It "kF	str		Sent by scroll-forward/down key."
.It "kH	str		Sent by home-down key."
.It "kh	str		Sent by home key."
.It "kI	str		Sent by insert-character or enter-insert-mode key."
.It "kL	str		Sent by delete-line key."
.It "kl	str		Sent by left-arrow key."
.It "kM	str		Sent by insert key while in insert mode."
.It "km	bool		Has a"
.Dq meta
key (shift, sets parity bit).
.It "kN	str		Sent by next-page key."
.It "kn	num	(o)	Number of function"
.Pq Sy \&k\&0 Ns \- Ns Sy \&k\&9
keys (default 0).
.It "ko	str	(o)	Termcap entries for other non-function keys."
.It "kP	str		Sent by previous-page key."
.It "kR	str		Sent by scroll-backward/up key."
.It "kr	str		Sent by right-arrow key."
.It "kS	str		Sent by clear-to-end-of-screen key."
.It "ks	str		Put terminal in"
.Dq keypad transmit
mode.
.It "kT	str		Sent by set-tab key."
.It "kt	str		Sent by clear-tab key."
.It "ku	str		Sent by up-arrow key."
.It "l0-l9	str		Labels on function keys if not"
.Dq \&f Ns Em n .
.It "LC	bool	(o)	Lower-case only."
.It "LE	str	(NP)	Move cursor left"
.Ar n
positions.
.It "le	str	(P)	Move cursor left one position."
.It "li	num		Number of lines on screen or page (see"
.Sx BUGS
section below).
.It "ll	str		Last line, first column."
.It "lm	num		Lines of memory if >" Sy \&li
(0 means varies).
.It "ma	str	(o)	Arrow key map (used by"
.Xr \&vi
version 2 only).
.It "mb	str		Turn on blinking attribute."
.It "md	str		Turn on bold (extra bright) attribute."
.It "me	str		Turn off all attributes."
.It "mh	str		Turn on half-bright attribute."
.It "mi	bool		Safe to move while in insert mode."
.It "mk	str		Turn on blank attribute (characters invisible)."
.It "ml	str	(o)	Memory lock on above cursor."
.It "mm	str		Turn on"
.Dq meta mode
(8th bit).
.It "mo	str		Turn off"
.Dq meta mode .
.It "mp	str		Turn on protected attribute."
.It "mr	str		Turn on reverse-video attribute."
.It "ms	bool		Safe to move in standout modes."
.It "mu	str	(o)	Memory unlock (turn off memory lock)."
.It "nc	bool	(o)	No correctly-working"
.Sy \&cr
(Datamedia 2500, Hazeltine 2000).
.It "nd	str		Non-destructive space (cursor right)."
.It "NL	bool	(o)" Ta Sy \&\en No "is newline, not line feed."
.It "nl	str	(o)	Newline character if not" Sy \en .
.It "ns	bool	(o)	Terminal is a" Tn CRT No "but doesn't scroll."
.It "nw	str	(P)	Newline (behaves like"
.Sy \&cr
followed by
.Sy \&do ) .
.It "OP	bool	(o)	Odd parity."
.It "os	bool		Terminal overstrikes."
.It "pb	num		Lowest baud where delays are required."
.It "pc	str		Pad character (default" Tn NUL ).
.It "pf	str		Turn off the printer."
.It "pk	str		Program function key"
.Em n
to type string
.Em s
.Pf ( Xr terminfo
only).
.It "pl	str		Program function key"
.Em n
to execute string
.Em s
.Pf ( Xr terminfo
only).
.It "pO	str	(N)	Turn on the printer for"
.Em n
bytes.
.It "po	str		Turn on the printer."
.It "ps	str		Print contents of the screen."
.It "pt	bool	(o)	Has hardware tabs (may need to be set with"
.Sy \&is ) .
.It "px	str		Program function key"
.Em n
to transmit string
.Em s
.Pf ( Xr terminfo
only).
.It "r1-r3	str		Reset terminal completely to sane modes"
.Pf ( Xr terminfo
only).
.It "rc	str	(P)	Restore cursor to position of last"
.Sy \&sc .
.It "rf	str		Name of file containing reset codes."
.It "RI	str	(NP)	Move cursor right"
.Em n
positions.
.It "rp	str	(NP*)	Repeat character"
.Em c n
times.
.It "rs	str		Reset terminal completely to sane modes"
.Pf ( Nm termcap
only).
.It "sa	str	(NP)	Define the video attributes."
.It "sc	str	(P)	Save cursor position."
.It "se	str		End standout mode."
.It "SF	str	(NP*)	Scroll forward"
.Em n
lines.
.It "sf	str	(P)	Scroll text up."
.It "sg	num		Number of garbage chars left by"
.Sy \&so
or
.Sy \&se
(default 0).
.It "so	str		Begin standout mode."
.It "SR	str	(NP*)	Scroll backward"
.Em n
lines.
.It "sr	str	(P)	Scroll text down."
.It "st	str		Set a tab in all rows, current column."
.It "ta	str	(P)	Tab to next 8-position hardware tab stop."
.It "tc	str		Entry of similar terminal; must be last."
.It "te	str		String to end programs that use"
.Nm termcap .
.It "ti	str		String to begin programs that use"
.Nm termcap .
.It "ts	str	(N)	Go to status line, column"
.Em n .
.It "UC	bool	(o)	Upper-case only."
.It "uc	str		Underscore one character and move past it."
.It "ue	str		End underscore mode."
.It "ug	num		Number of garbage chars left by"
.Sy \&us
or
.Sy \&ue
(default 0).
.It "ul	bool		Underline character overstrikes."
.It "UP	str	(NP*)	Move cursor up"
.Em n
lines.
.It "up	str		Upline (cursor up)."
.It "us	str		Start underscore mode."
.It "vb	str		Visible bell (must not move cursor)."
.It "ve	str		Make cursor appear normal (undo"
.Sy \&vs Ns / Ns Sy \&vi ) .
.It "vi	str		Make cursor invisible."
.It "vs	str		Make cursor very visible."
.It "vt	num		Virtual terminal number (not supported on all systems)."
.It "wi	str	(N)	Set current window."
.It "ws	num		Number of columns in status line."
.It "xb	bool		Beehive"
.Pf ( "f1=" Dv ESC ,
.Pf "f2=" Sy \&^C ) .
.It "xn	bool		Newline ignored after 80 columns (Concept)."
.It "xo	bool		Terminal uses xoff/xon"
.Pq Dv DC3 Ns / Ns Dv DC1
handshaking.
.It "xr	bool	(o)	Return acts like"
.Sy "ce cr nl"
(Delta Data).
.It "xs	bool		Standout not erased by overwriting (Hewlett-Packard)."
.It "xt	bool		Tabs ruin, magic"
.Sy \&so
char (Teleray 1061).

contrib/code/jupp/termcap.c
/* $MirOS: contrib/code/jupp/termcap.c,v 1.8 2012/12/30 21:45:17 tg Exp $ */
/*
 *	TERMCAP/TERMINFO database interface
 *	Copyright
 *		(C) 1992 Joseph H. Allen
 *
 *	This file is part of JOE (Joe's Own Editor)
 */
#include "config.h"
#include "types.h"

#include <stdio.h>
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef TERMINFO
#ifdef HAVE_TERM_H
#endif
#endif

#include "blocks.h"
#include "path.h"
#include "termcap.h"
#include "utils.h"
#include "va.h"
#include "vs.h"

int dopadding = 0;
unsigned char *joeterm = NULL;

/* Default termcap entry */

unsigned char defentry[] = "\
:co#80:li#25:am:\
:ho=\\E[H:cm=\\E[%i%d;%dH:cV=\\E[%i%dH:\
:up=\\E[A:UP=\\E[%dA:DO=\\E[%dB:nd=\\E[C:RI=\\E[%dC:LE=\\E[%dD:\
:cd=\\E[J:ce=\\E[K:cl=\\E[H\\E[J:\
:so=\\E[7m:se=\\E[m:us=\\E[4m:ue=\\E[m:\
:mb=\\E[5m:md=\\E[1m:mh=\\E[2m:me=\\E[m:\
:ku=\\E[A:kd=\\E[B:kl=\\E[D:kr=\\E[C:\
:al=\\E[L:AL=\\E[%dL:dl=\\E[M:DL=\\E[%dM:\
:ic=\\E[@:IC=\\E[%d@:dc=\\E[P:DC=\\E[%dP:\
";

/* Return true if termcap line matches name */

static int match(unsigned char *s, unsigned char *name)
{
	if (s[0] == 0 || s[0] == '#')
		return 0;
	do {
		int x;

		for (x = 0; s[x] == name[x] && name[x] && s[x]; ++x) ;
		if (name[x] == 0 && (s[x] == ':' || s[x] == '|'))
			return 1;
		while (s[x] != ':' && s[x] != '|' && s[x])
			++x;
		s += x + 1;
	} while (s[-1] == '|');
	return 0;
}

/* Find termcap entry in a file */

static unsigned char *lfind(unsigned char *s, int pos, FILE *fd, unsigned char *name)
{
	int c, x;

	if (!s)
		s = vsmk(1024);
      loop:
	while (c = getc(fd), c == ' ' || c == '\t' || c == '#')
		do {
			c = getc(fd);
		} while (!(c == -1 || c == '\n'));
	if (c == -1)
		return s = vstrunc(s, pos);
	ungetc(c, fd);
	s = vstrunc(s, x = pos);
	while (1) {
		c = getc(fd);
		if (c == -1 || c == '\n')
			if (x != pos && s[x - 1] == '\\') {
				--x;
				if (!match(s + pos, name))
					goto loop;
				else
					break;
			} else if (!match(s + pos, name))
				goto loop;
			else
				return vstrunc(s, x);
		else if (c == '\r')
			/* do nothing */;
		else {
			s = vsset(s, x, c);
			++x;
		}
	}
	while (c = getc(fd), c != -1)
		if (c == '\n')
			if (s[x - 1] == '\\')
				--x;
			else
				break;
		else if (c == '\r')
			/* do nothing */;
		else {
			s = vsset(s, x, c);
			++x;
		}
	s = vstrunc(s, x);
	return s;
}

/* Lookup termcap entry in index */

static off_t findidx(FILE *file, unsigned char *name)
{
	unsigned char buf[80];
	off_t addr = 0;

	while (fgets((char *)buf, 80, file)) {
		int x = 0, flg = 0, c, y, z;

		do {
			for (y = x; buf[y] && buf[y] != ' ' && buf[y] != '\n'; ++y) ;
			c = buf[y];
			buf[y] = 0;
			if (c == '\n' || !c) {
				z = 0;
				sscanf((char *)(buf + x), "%x", &z);
				addr += z;
			} else if (!strcmp(buf + x, name))
				flg = 1;
			x = y + 1;
		} while (c && c != '\n');
		if (flg)
			return addr;
	}
	return 0;
}

/* Load termcap entry */

CAP *getcap(unsigned char *name, unsigned int baud, void (*out) (unsigned char *, unsigned char), void *outptr)
{
	CAP *cap;
	FILE *f, *f1;
	off_t idx;
	int x, y, c, z, ti;
	unsigned char *tp, *pp, *qq, *namebuf, **npbuf, *idxname;
	int sortsiz;

	if (!name && !(name = joeterm) && !(name = (unsigned char *)getenv("TERM")))
		return NULL;
	cap = (CAP *) joe_malloc(sizeof(CAP));
	cap->tbuf = vsmk(4096);
	cap->abuf = NULL;
	cap->sort = NULL;
	cap->paste_on = NULL;
	cap->paste_off = NULL;

	if (!strcmp(name, "xterm-xfree86")) {
		cap->paste_on = "\033[?2004h";
		cap->paste_off = "\033[?2004l";
	}

#ifdef TERMINFO
	cap->abuf = (unsigned char *) joe_malloc(4096);
	cap->abufp = cap->abuf;
	if (tgetent((char *)cap->tbuf, (char *)name) == 1)
		return setcap(cap, baud, out, outptr);
	else {
		joe_free(cap->abuf);
		cap->abuf = NULL;
	}
#endif

	name = vsncpy(NULL, 0, sz(name));
	cap->sort = (struct sortentry *) joe_malloc(sizeof(struct sortentry) * (sortsiz = 64));

	cap->sortlen = 0;

	tp = (unsigned char *)getenv("TERMCAP");

	if (tp && tp[0] == '/')
		namebuf = vsncpy(NULL, 0, sz(tp));
	else {
		if (tp)
			cap->tbuf = vsncpy(sv(cap->tbuf), sz(tp));
		if ((tp = (unsigned char *)getenv("TERMPATH")))
			namebuf = vsncpy(NULL, 0, sz(tp));
		else {
			if ((tp = (unsigned char *)getenv("HOME"))) {
				namebuf = vsncpy(NULL, 0, sz(tp));
				namebuf = vsadd(namebuf, '/');
			} else
				namebuf = NULL;
			namebuf = vsncpy(sv(namebuf), sc(".termcap "));
			if (has_JOERC &&
			    vsscan(sz(get_JOERC), sc("\t :")) == ~0) {
				namebuf = vsncpy(sv(namebuf), sz(get_JOERC));
				namebuf = vsncpy(sv(namebuf), sc("termcap "));
			}
			namebuf = vsncpy(sv(namebuf), sc("/etc/termcap"));
		}
	}

	npbuf = vawords(NULL, sv(namebuf), sc("\t :"));
	vsrm(namebuf);

	y = 0;
	ti = 0;

	if (match(cap->tbuf, name))
		goto checktc;

	cap->tbuf = vstrunc(cap->tbuf, 0);

      nextfile:
	if (!npbuf[y]) {
/*
 varm(npbuf);
 vsrm(name);
 vsrm(cap->tbuf);
 joe_free(cap->sort);
 joe_free(cap);
 return 0;
*/
		fprintf(stderr, "Couldn't load termcap entry.  Using ansi default\n");
		ti = 0;
		cap->tbuf = vsncpy(cap->tbuf, 0, sc(defentry));
		goto checktc;
	}
	idx = 0;
	idxname = vsncpy(NULL, 0, sz(npbuf[y]));
	idxname = vsncpy(idxname, sLEN(idxname), sc(".idx"));
	f1 = fopen((char *)(npbuf[y]), "r");
	++y;
	if (!f1)
		goto nextfile;
	f = fopen((char *)idxname, "r");
	if (f) {
		struct stat buf, buf1;

		fstat(fileno(f), &buf);
		fstat(fileno(f1), &buf1);
		if (buf.st_mtime > buf1.st_mtime)
			idx = findidx(f, name);
		else
			fprintf(stderr, "%s is out of date\n", idxname);
		fclose(f);
	}
	vsrm(idxname);
#ifdef HAVE_FSEEKO
	fseeko(f1, idx, 0);
#else
	/* ugh. SOL! */
	fseek(f1, (long)idx, 0);
#endif
	cap->tbuf = lfind(cap->tbuf, ti, f1, name);
	fclose(f1);
	if (sLEN(cap->tbuf) == ti)
		goto nextfile;

      checktc:
	x = sLEN(cap->tbuf);
	do {
		cap->tbuf[x] = 0;
		while (x && cap->tbuf[--x] != ':')
			/* do nothing */;
	} while (x && (!cap->tbuf[x + 1] || cap->tbuf[x + 1] == ':'));

	if (cap->tbuf[x + 1] == 't' && cap->tbuf[x + 2] == 'c' && cap->tbuf[x + 3] == '=') {
		name = vsncpy(NULL, 0, sz(cap->tbuf + x + 4));
		cap->tbuf[x] = 0;
		cap->tbuf[x + 1] = 0;
		ti = x + 1;
		sLen(cap->tbuf) = x + 1;
		if (y)
			--y;
		goto nextfile;
	}

      doline:
	pp = cap->tbuf + ti;

/* Process line at pp */

      loop:
	while (*pp && *pp != ':')
		++pp;
	if (*pp) {
		int q;

		*pp++ = 0;
	      loop1:
		if (pp[0] == ' ' || pp[0] == '\t')
			goto loop;
		for (q = 0; pp[q] && pp[q] != '#' && pp[q] != '=' && pp[q] != '@' && pp[q] != ':'; ++q) ;
		qq = pp;
		c = pp[q];
		pp[q] = 0;
		if (c)
			pp += q + 1;
		else
			pp += q;

		x = 0;
		y = cap->sortlen;
		z = -1;
		if (!y) {
			/* dead store: z = 0; */
			goto in;
		}
		while (z != (x + y) / 2) {
			int found;

			z = (x + y) / 2;
			found = strcmp(qq, cap->sort[z].name);
			if(found > 0) {
				x = z;
			} else if(found < 0) {
				y = z;
			} else {
				if (c == '@')
					mmove(cap->sort + z, cap->sort + z + 1, (cap->sortlen-- - (z + 1)) * sizeof(struct sortentry));

				else if (c && c != ':')
					cap->sort[z].value = qq + q + 1;
				else
					cap->sort[z].value = NULL;
				if (c == ':')
					goto loop1;
				else
					goto loop;
			}
		}
	      in:
		if (cap->sortlen == sortsiz)
			cap->sort = (struct sortentry *) joe_realloc(cap->sort, (sortsiz += 32) * sizeof(struct sortentry));
		mmove(cap->sort + y + 1, cap->sort + y, (cap->sortlen++ - y) * sizeof(struct sortentry));

		cap->sort[y].name = qq;
		if (c && c != ':')
			cap->sort[y].value = qq + q + 1;
		else
			cap->sort[y].value = NULL;
		if (c == ':')
			goto loop1;
		else
			goto loop;
	}

	if (ti) {
		for (--ti; ti; --ti)
			if (!cap->tbuf[ti - 1])
				break;
		goto doline;
	}

	varm(npbuf);
	vsrm(name);

	cap->pad = jgetstr(cap, US "pc");
	if (dopadding)
		cap->dopadding = 1;
	else
		cap->dopadding = 0;

/* show sorted entries
	for(x=0;x!=cap->sortlen;++x)
		printf("%s = %s\n",cap->sort[x].name,cap->sort[x].value);
*/
	return setcap(cap, baud, out, outptr);
}

static struct sortentry *findcap(CAP *cap, unsigned char *name)
{
	int x, y, z;
	int found;

	x = 0;
	y = cap->sortlen;
	z = -1;
	while (z != (x + y) / 2) {
		z = (x + y) / 2;
		found = strcmp(name, cap->sort[z].name);
		if (found > 0)
			x = z;
		else if (found < 0)
			y = z;
		else
			return cap->sort + z;
	}
	return NULL;
}

CAP *setcap(CAP *cap, unsigned int baud, void (*out) (unsigned char *, unsigned char), void *outptr)
{
	cap->baud = baud;
	cap->div = 100000 / baud;
	cap->out = out;
	cap->outptr = outptr;
	return cap;
}

int getflag(CAP *cap, unsigned char *name)
{
#ifdef TERMINFO
	if (cap->abuf)
		return tgetflag((char *)name);
#endif
	return findcap(cap, name) != NULL;
}

unsigned char *jgetstr(CAP *cap, unsigned char *name)
{
	struct sortentry *s;

#ifdef TERMINFO
	if (cap->abuf)
		return (unsigned char *)tgetstr((char *)name, (char **)&cap->abufp);
#endif
	s = findcap(cap, name);
	if (s)
		return s->value;
	else
		return NULL;
}

int getnum(CAP *cap, unsigned char *name)
{
	struct sortentry *s;

#ifdef TERMINFO
	if (cap->abuf)
		return tgetnum((char *)name);
#endif
	s = findcap(cap, name);
	if (s && s->value)
		return atoi((char *)(s->value));
	return -1;
}

void rmcap(CAP *cap)
{
	vsrm(cap->tbuf);
	if (cap->abuf)
		joe_free(cap->abuf);
	if (cap->sort)
		joe_free(cap->sort);
	joe_free(cap);
}

static unsigned char escape(unsigned char **s)
{
	unsigned char c = *(*s)++;

	if (c == '^' && **s)
		if (**s != '?')
			return 037 & *(*s)++;
		else {
			(*s)++;
			return 127;
		}
	else if (c == '\\' && **s)
		switch (c = *((*s)++)) {
		case '0':
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
			c -= '0';
			if (**s >= '0' && **s <= '7')
				c = (c << 3) + *((*s)++) - '0';
			if (**s >= '0' && **s <= '7')
				c = (c << 3) + *((*s)++) - '0';
			return c;
		case 'e':
		case 'E':
			return 27;
		case 'n':
		case 'l':
			return 10;
		case 'r':
			return 13;
		case 't':
			return 9;
		case 'b':
			return 8;
		case 'f':
			return 12;
		case 's':
			return 32;
		default:
			return c;
	} else
		return c;
}

#ifdef TERMINFO
static CAP *outcap;
static int outout(int c)
{
	outcap->out(outcap->outptr, c);
	return(c);	/* act like putchar() - return written char */
}
#endif

void texec(CAP *cap, unsigned char *s, int l, int a0, int a1, int a2, int a3)
{
	int c, tenth = 0, x;
	int args[4];
	int vars[128];
	int *a = args;

/* Do nothing if there is no string */
	if (!s)
		return;

#ifdef TERMINFO
	if (cap->abuf) {
		unsigned char *aa;

		outcap = cap;
		aa = (unsigned char *)tgoto((char *)s, a1, a0);
		tputs((char *)aa, l, outout);
		return;
	}
#endif

/* Copy args into array (yuk) */
	args[0] = a0;
	args[1] = a1;
	args[2] = a2;
	args[3] = a3;

/* Get tenths of MS of padding needed */
	while (*s >= '0' && *s <= '9')
		tenth = tenth * 10 + *s++ - '0';
	tenth *= 10;
	if (*s == '.') {
		++s;
		tenth += *s++ - '0';
	}

/* Check if we have to multiply by number of lines */
	if (*s == '*') {
		++s;
		tenth *= l;
	}

/* Output string */
	while ((c = *s++) != '\0')
		if (c == '%' && *s) {
			switch (x = a[0], c = escape(&s)) {
			case 'C':
				if (x >= 96) {
					cap->out(cap->outptr, x / 96);
					x %= 96;
				}
			case '+':
				if (*s)
					x += escape(&s);
			case '.':
				cap->out(cap->outptr, x);
				++a;
				break;
			case 'd':
				if (x < 10)
					goto one;
			case '2':
				if (x < 100)
					goto two;
			case '3':
				c = '0';
				while (x >= 100) {
					++c;
					x -= 100;
				}
				cap->out(cap->outptr, c);
			      two:c = '0';
				while (x >= 10) {
					++c;
					x -= 10;
				}
				cap->out(cap->outptr, c);
			      one:cap->out(cap->outptr, '0' + x);
				++a;
				break;
			case 'r':
				a[0] = a[1];
				a[1] = x;
				break;
			case 'i':
				++a[0];
				++a[1];
				break;
			case 'n':
				a[0] ^= 0140;
				a[1] ^= 0140;
				break;
			case 'm':
				a[0] ^= 0177;
				a[1] ^= 0177;
				break;
			case 'f':
				++a;
				break;
			case 'b':
				--a;
				break;
			case 'a':
				x = s[2];
				if (s[1] == 'p')
					x = a[x - 0100];
				switch (*s) {
				case '+':
					a[0] += x;
					break;
				case '-':
					a[0] -= x;
					break;
				case '*':
					a[0] *= x;
					break;
				case '/':
					a[0] /= x;
					break;
				case '%':
					a[0] %= x;
					break;
				case 'l':
					a[0] = vars[x];
					break;
				case 's':
					vars[x] = a[0];
					break;
				default:
					a[0] = x;
				}
				s += 3;
				break;
			case 'D':
				a[0] = a[0] - 2 * (a[0] & 15);
				break;
			case 'B':
				a[0] = 16 * (a[0] / 10) + a[0] % 10;
				break;
			case '>':
				if (a[0] > escape(&s))
					a[0] += escape(&s);
				else
					escape(&s);
			default:
				cap->out(cap->outptr, '%');
				cap->out(cap->outptr, c);
			}
		} else {
			--s;
			cap->out(cap->outptr, escape(&s));
		}

/* Output padding characters */
	if (cap->dopadding) {
		if (cap->pad)
			while (tenth >= cap->div)
				for (s = cap->pad; *s; ++s) {
					cap->out(cap->outptr, *s);
					tenth -= cap->div;
				}
		else
			while (tenth >= cap->div) {
				cap->out(cap->outptr, 0);
				tenth -= cap->div;
			}
	}
}

static int total;

static void cst(unsigned char *ptr, unsigned char c)
{
	++total;
}

int tcost(CAP *cap, unsigned char *s, int l, int a0, int a1, int a2, int a3)
{
	void (*out) (unsigned char *, unsigned char) = cap->out;

	if (!s)
		return 10000;
	total = 0;
	cap->out = cst;
	texec(cap, s, l, a0, a1, a2, a3);
	cap->out = out;
	return total;
}

static unsigned char *ssp;
static void cpl(unsigned char *ptr, unsigned char c)
{
	ssp = vsadd(ssp, c);
}

unsigned char *tcompile(CAP *cap, unsigned char *s, int a0, int a1, int a2, int a3)
{
	void (*out) (unsigned char *, unsigned char) = cap->out;
	int divider = cap->div;

	if (!s)
		return NULL;
	cap->out = cpl;
	cap->div = 10000;
	ssp = vsmk(10);
	texec(cap, s, 0, a0, a1, a2, a3);
	cap->out = out;
	cap->div = divider;
	return ssp;
}
